Add new ore preference system, to allow oredict outputs to use a specific mod if available

KnightMiner opened the pull request:
As originally discussed in #2884, this adds the feature to choose a preference for oredict outputs. For example, adding Immersive Engineering to the top of the list will cause IE copper and steel to be used for outputs over Thermal Expansion or Chisel blocks.

bonii-xx commented on src/main/java/slimeknights/tconstruct/common/config/Config.java:
```diff
@@ -46,6 +47,10 @@ private Config() {
   };
+  private static String[] orePreference = {
+      "minecraft",
+      "tconstruct"
```
Should probably put some more defaults here so people don't get chisel for copper etc. by default. TE always is a good fit ("thermalfoundation" I think), and immersive engineering.

bookerthegeek replied:
Industrial Craft would also be a good fit in this list

bonii-xx commented on src/main/java/slimeknights/tconstruct/library/MaterialIntegration.java:
```diff
@@ -145,14 +143,7 @@ public void integrate() {
   private void registerRepresentativeItem() {
     // also set the representative item
     if(material.getRepresentativeItem().isEmpty() && representativeItem != null && !representativeItem.isEmpty()) {
-      List<ItemStack> ore = OreDictionary.getOres(representativeItem, false);
```
Wondering what the impact of existing mods is for that. Like XU2 that adds materials. Or IE.

KnightMiner replied:
With the way I set it up, if they manually set the item there is no change since thats what the old code did with the ore names. If they use an oredict name will be the only changes, but they did not have control over that before anyways

bonii-xx replied:
Ok, great

bonii-xx commented on src/main/java/slimeknights/tconstruct/library/smeltery/OreCastingRecipe.java:
```diff
@@ -4,42 +4,56 @@
 public class OreCastingRecipe extends CastingRecipe {
 
+  /** @Deprecated use oreName instead */
+  @Deprecated
```
I wonder if it wouldn't be better if we made a new class for this, and deprecated the old class. Seems a bit cleaner, but less compatible (although I hope nobody casts to that class without checking for instance first..)

KnightMiner replied:
I considered that, though I was not sure what to name it.

bonii-xx replied:
OreCastingRecipe2, OreCastingRecipeNew, BetterOreCastingRecipe, UseThisInsteadOfTheOldOneOreCastingRecipe, the terrible possibilities are endless!
I guess yours or maybe SmartOreCastingRecipe or possibly even just OreDictCastingRecipe.. hrm.

bonii-xx commented on src/main/java/slimeknights/tconstruct/library/utils/RecipeUtil.java:
```diff
@@ -0,0 +1,86 @@
+import net.minecraftforge.oredict.OreDictionary;
+import slimeknights.mantle.client.CreativeTab;
+
+public class RecipeUtil {
```
final

bonii-xx commented on src/main/java/slimeknights/tconstruct/library/utils/RecipeUtil.java:
```diff
@@ -0,0 +1,86 @@
+    for(String mod : orePreferences) {
+      Optional<ItemStack> optional = items
+          .stream()
+          .filter(stack -> !stack.isEmpty() && stack.getItem().getRegistryName().getResourceDomain().equals(mod))
```
anyMatch()
btw I can strongly recommend https://medium.com/@tagir_valeev/use-stream-api-simpler-or-dont-use-it-at-all-ea0a44a4b1ff
Very useful for stream stuff

bonii-xx commented on src/main/java/slimeknights/tconstruct/library/utils/RecipeUtil.java:
```diff
@@ -0,0 +1,86 @@
+  public static ItemStack getPreference(String oreName) {
+    // if we found one before, use that
+    if(preferenceCache.containsKey(oreName)) {
+      return preferenceCache.get(oreName).copy();
```
Very good

KnightMiner replied:
We used that in some of the model stuff, didn't we. Only downside is we would have to cache if we got no result, if that is acceptable

bonii-xx replied:
No, the remaining code would have to be adjusted to fit a single call, but that's just some code reordering. You already cache the first oredict entry in your code, it'd just require to move all the remaining code (aka get oredict items) into the cache method.

KnightMiner replied:
Looking into It, I guess I can just return null from thay function if the item does not exist (so no cache) then use Map.getOrDefault to return ItemStack.EMPTY if the result is null.

bonii-xx commented on src/main/java/slimeknights/tconstruct/library/utils/RecipeUtil.java:
```diff
@@ -0,0 +1,86 @@
+    // ensure we do not return a wildcard value
+    if(preference.getMetadata() == OreDictionary.WILDCARD_VALUE) {
+      NonNullList<ItemStack> subItems = NonNullList.create();
+      preference.getItem().getSubItems(CreativeTab.SEARCH, subItems);
```
Is that Serverside-Safe?

KnightMiner replied:
Yes, it is now since one of the 1.12 builds. Might need to double check if that was 1.12.1 only though...

bonii-xx replied:
Nice. I'm not up to date with 1.12+ changes.

bonii-xx commented:
Looks good overall, just a few details I'm pondering on. I'll leave them to your discretion.

KnightMiner commented:
Pushed some commits addressing the concerns

While testing after making some of these, I noticed an issue with our JEI integration which relates to this. It seems because of the way JEI now handles validations (on load instead of when viewing recipes), our empty OreCastingRecipes still show (and by extendsion empty PreferenceCastingRecipes).

According to mezz, it should be safe to just not load those recipes if they are empty by the time JEI loads as the oredict should be stable by then, though if I do that would it make sense to also cache the empty results from the preferences or leave those uncached?
